Defect/minor cleanup and 9.0 release

pkoutoupis opened the pull request:
Cleaned up docs/some code formatting, fixed spec for sles, updated changelog, copyright year and version to 9.0.0. Also fixed resize operation error messaging reported in issue #142. Even fixed and cleaned up NVMe Target code while adding comments for clarity.

pkoutoupis commented:
@matteotenca Also, confirming: we will hold off for 9.1.0 to release your rapiddisk-on-boot updates? If so, I will immediately cut the 9.0.0 tag release (after the merge). Let me know otherwise.

matteotenca commented:
> @matteotenca Also, confirming: we will hold off for 9.1.0 to release your rapiddisk-on-boot updates? If so, I will immediately cut the 9.0.0 tag release (after the merge). Let me know otherwise.

Let's wait 9.1.0 for the new `rapiddisk-on-boot`, I need some time to improve and fix the script yet!

matteotenca commented on doc/rapiddiskd.1:
```diff
@@ -7,14 +7,17 @@ rapiddiskd [ -h | -v ] [ options ]
 .TP
 -V
-Enable debug messages to stdout (this is ugly)
+Enable debug messages to stdout (this is ugly).
```
This should be changed to `stderr` too.

pkoutoupis replied:
Will do.

matteotenca commented on src/utils.c:
```diff
@@ -1,6 +1,9 @@
 /**
+ * @file util.c
```
`util.c` should be changed to `utils.c`

pkoutoupis replied:
Will do.

matteotenca reviewed (changes requested):
Just a couple of ultra small issues.

matteotenca commented on src/nvmet.c:
```diff
@@ -905,36 +881,31 @@ int nvmet_unexport_volume(char *device, char *host, int port, char *return_messa
-			goto host_check_out;
-		}
+	}
+
+	/*
+	 * Grab number of entries in allowed_hosts directory. If more than zero still exist, we will not
+	 * proceed with the rest
+	 */
+	sprintf(path, "%s/%s%s-%s/allowed_hosts/", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
+	if ((err = scandir(path, &list, NULL, NULL)) < 0) {
+		msg = "Error. Unable to scan host exports directory for %s. %s: unlink: %s";
+		print_error(msg, return_message, path, __func__, strerror(errno));
+		return rc;
+	}
+	hostno = err;
```
I believe there is a problem here. `err` will be always >= 2, because `.` and `..` are always present.

At line 896 `hostno` is set the value of `err`, so is always > 0.
Every subsequent checks that require `hostno` to be == 0, will fail. This leads to skip some phases of the unexporting process.

Check at line 908 will always fail. The same happens at line 934.

Here an example of what happens:

```bash
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        None.


Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -N
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Exported NVMe Ports

        1: Port: 1000 - 192.168.20.40 (tcp)
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -g -j -e -b rd2 -P 1000 -H nqn.host1 | jq
{
  "status": "Success",
  "message": "Block device rd2 has been mapped to nqn.host1 through port 1000 as nqn.2021-06.org.rapiddisk:ubuserver-rd2."
}
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        1: NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd2         Namespace: 1    Device: /dev/rd2        Enabled: True

Exported NVMe Ports

        1: Port: 1000 - 192.168.20.40 (tcp)     NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd2
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -g -j -x -b rd2 -P 1000 -H nqn.host1 | jq
{
  "status": "Success"
}
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        1: NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd2         Namespace: 1    Device: /dev/rd2        Enabled: True

Exported NVMe Ports

        1: Port: 1000 - 192.168.20.40 (tcp)     NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd2
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -N
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Exported NVMe Ports

        1: Port: 1000 - 192.168.20.40 (tcp)
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$
```

I tried implementing a `scandir` filter like this, which causes `..` and `.` to be removed from the result passed to `scandir`:

```C
int scandir_filter_no_dot(const struct dirent *list) {
	if ((strncmp(list->d_name, "..", 2) == SUCCESS) || (strncmp(list->d_name, ".", 1) == SUCCESS)) {
		return FALSE;
	}
	return TRUE;
}
```

Passing it as `selector` to the scandir call at line 891, cause `err` to be 0 if nothing except `..` and `.` is found in the path. This approach can lead to avoid the loop at line 897, and gives a chance to checks at lines 908 and 934 to succeed.

Maybe I am missing something, your knowledge is needed here!

pkoutoupis replied:
Wow. I completely overlooked `.` and `..`. I do not know what I was thinking. Nice catch. Looking at your examples, it would be more useful to see it without the `-j`.

I imagine that in your example, because you specified a -H, it removed the mapping from to the hostnqn and did not remove the target or namespace from the filesystem.

Let me look into addressing this and I will probably need to write out a document outlining expectations with command execution.

pkoutoupis replied:
The goal is to make sure that there are no more connected hosts before ultimately removing the port mapping, namespace and target from the subsystem. So, I should have done this from the beginning:
```
diff --git a/src/nvmet.c b/src/nvmet.c
index 533bcae..2792dc5 100644
--- a/src/nvmet.c
+++ b/src/nvmet.c
@@ -891,10 +891,10 @@ int nvmet_unexport_volume(char *device, char *host, int port, char *return_messa
                print_error(msg, return_message, path, __func__, strerror(errno));
                return rc;
        }
-       hostno = err;
        if (err > 0) {
                for (n = 0; n < err; n++) {
                        if (strncmp(list[n]->d_name, ".", 1) != SUCCESS) {
+                               hostno++;
                                msg = "The target is still mapped to host NQN %s.";
                                print_error(msg, return_message, list[n]->d_name);
                        }
```
That way we avoid counting anything else in the `allowed_hosts` directory.

But I still do need to document this thoroughly to show expectations when certain arguments and conditions are present during command execution.

matteotenca replied:
This is another approach:

```diff
diff --git a/src/nvmet.c b/src/nvmet.c
index f91e3ab..a95abf6 100644
--- a/src/nvmet.c
+++ b/src/nvmet.c
@@ -802,6 +802,20 @@ int nvmet_revalidate_size(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, char
        return SUCCESS;
 }

+/**
+ * > If the file name starts with a dot, return false. Otherwise, return true
+ *
+ * @param list This is the directory entry that is being passed to the function.
+ *
+ * @return the number of files in the directory.
+ */
+int scandir_filter_no_dot(const struct dirent *list) {
+       if (strncmp(list->d_name, ".", 1) == SUCCESS) {
+               return FALSE;
+       }
+       return TRUE;
+}
+
 /*
  * description: Unexport an NVMe Target.
  */
@@ -886,19 +900,16 @@ int nvmet_unexport_volume(char *device, char *host, int port, char *return_messa
         * proceed with the rest
         */
        sprintf(path, "%s/%s%s-%s/allowed_hosts/", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
-       if ((err = scandir(path, &list, NULL, NULL)) < 0) {
+       if ((err = scandir(path, &list, scandir_filter_no_dot, NULL)) < 0) {
                msg = "Error. Unable to scan host exports directory for %s. %s: unlink: %s";
                print_error(msg, return_message, path, __func__, strerror(errno));
                return rc;
        }
-       hostno = err;
-       if (err > 0) {
-               for (n = 0; n < err; n++) {
-                       if (strncmp(list[n]->d_name, ".", 1) != SUCCESS) {
-                               msg = "The target is still mapped to host NQN %s.";
-                               print_error(msg, return_message, list[n]->d_name);
-                       }
-               }
+
+       for (n = 0; n < err; n++) {
+               hostno++;
+               msg = "The target is still mapped to host NQN %s.";
+               print_error(msg, return_message, list[n]->d_name);
        }
        list = clean_scandir(list, err);

```
```

pkoutoupis replied:
That is fine: I can add the `scandir_filter_no_dot` but I may add it in another file like the utils file because it can and may be used elsewhere in the future. I will push new code today.

pkoutoupis commented:
@matteotenca Let me know what you think with the latest push. Also documented NVMe target functionality in both the manpage and the README.

pkoutoupis commented:
@matteotenca Just following up. I am hoping to tag release 9.0.0 soon because I wanted to work on a few things for 9.1.0 with your upcoming rapiddisk-on-boot work.

matteotenca commented on src/sys.h:
```diff
@@ -37,7 +37,7 @@ SPDX-License-Identifier: GPL-2.0-or-later
 #endif
-int get_memory_usage(struct MEM_PROFILE *, char *return_message);
-struct VOLUME_PROFILE *search_volumes_targets(char *return_message);
+int get_memory_usage(struct MEM_PROFILE *, char *);
```
I believe that leaving the variables' names is useful, at least with my IDE (JetBrains). But it is just cosmetic...

pkoutoupis replied:
I will change it to add the var names.

matteotenca commented on src/utils.h:
```diff
@@ -36,6 +36,7 @@ SPDX-License-Identifier: GPL-2.0-or-later
 
 int preg_replace(const char *re, char *replacement, char *subject, char *result, size_t pcre2_result_len);
 int split(char* input_string, char** output_arr, char* delim);
+int scandir_filter_no_dot(const struct dirent *);
```
Same consideration about leaving vars names in declarations.

pkoutoupis replied:
I will change it to add the var names.

matteotenca commented:
I found a strange behaviour. In a comment, you wrote:

https://github.com/pkoutoupis/rapiddisk/blob/1f668284f5b9341241268d39d13137475fa65934/src/nvmet.c#L543-L546

But this allows not only to add new allowed hosts, but also to re-add the same one:

```bash
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -e -b rd0 -P 1 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been mapped to test through port 1 as nqn.2021-06.org.rapiddisk:ubuserver-rd0.
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -e -b rd0 -P 1 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been mapped to test through port 1 as nqn.2021-06.org.rapiddisk:ubuserver-rd0.
shub@ubuserver:~/rapiddisk/src$
```

And:

```bash
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -e -b rd0 -P 1
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been mapped to all hosts through port 1 as nqn.2021-06.org.rapiddisk:ubuserver-rd0.
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -e -b rd0 -P 1
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been mapped to all hosts through port 1 as nqn.2021-06.org.rapiddisk:ubuserver-rd0.
shub@ubuserver:~/rapiddisk
```

Maybe this is the wanted behaviour.

Furthermore, when exporting a target without specifying any allowed host, a subsequent invocation to export the same target but specifying an allowed host will fail, but with a generic error caused by the `symlink` call failing, not with a handled exception:

```bash
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -N
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Exported NVMe Ports

        1: Port: 1 - 192.168.20.40 (tcp)
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        None.


Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -e -b rd0 -P 1
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been mapped to all hosts through port 1 as nqn.2021-06.org.rapiddisk:ubuserver-rd0.
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -e -b rd0 -P 1 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Error. Unable to link host to port. nvmet_export_volume: symlink: Invalid argument
shub@ubuserver:~/rapiddisk/src$
```

With a modified error message:

```bash
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -x -b rd0 -P 1
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Error. NQN export: nqn.2021-06.org.rapiddisk:ubuserver-rd0 does not exist.
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -e -b rd0 -P 1
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been mapped to all hosts through port 1 as nqn.2021-06.org.rapiddisk:ubuserver-rd0.
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -e -b rd0 -P 1 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Error. Unable to link host to port. nvmet_export_volume: symlink: Invalid argument - from: /sys/kernel/config/nvmet/hosts/test, to: /sys/kernel/config/nvmet/subsystems/nqn.2021-06.org.rapiddisk:ubuserver-rd0/allowed_hosts/test
shub@ubuserver:~/rapiddisk/src$
```

The lines involved, especially line 621.

https://github.com/pkoutoupis/rapiddisk/blob/1f668284f5b9341241268d39d13137475fa65934/src/nvmet.c#L581-L627

Another issue (maybe): some dirs in the kernel filesystem are not removed properly (I am not sure):

```bash
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        None.


Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -e -b rd0 -P 1 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been mapped to test through port 1 as nqn.2021-06.org.rapiddisk:ubuserver-rd0.
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -e -b rd0 -P 1 -H test2
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been mapped to test2 through port 1 as nqn.2021-06.org.rapiddisk:ubuserver-rd0.
shub@ubuserver:~/rapiddisk/src$ ll /sys/kernel/config/nvmet/hosts/
total 0
drwxr-xr-x 4 root root 0 Jan  7 01:06 ./
drwxr-xr-x 5 root root 0 Jan  7 00:12 ../
drwxr-xr-x 2 root root 0 Jan  7 01:06 test/
drwxr-xr-x 2 root root 0 Jan  7 01:06 test2/
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -x -b rd0 -P 1 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been unmapped from NVMe Target host test.
The target is still mapped to host NQN test2.
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -x -b rd0 -P 1 -H test2
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been unmapped from NVMe Target host test2.
Block device rd0 has been unmapped from NVMe Target port 1.
Block device rd0 has been removed from the NVMe Target subsystem.
shub@ubuserver:~/rapiddisk/src$ ll /sys/kernel/config/nvmet/hosts/
total 0
drwxr-xr-x 4 root root 0 Jan  7 01:06 ./
drwxr-xr-x 5 root root 0 Jan  7 00:12 ../
drwxr-xr-x 2 root root 0 Jan  7 01:06 test/
drwxr-xr-x 2 root root 0 Jan  7 01:06 test2/
shub@ubuserver:~/rapiddisk/src$
```

It seems that upon unexporting, only the symlinks are deleted, not the dirs into `/sys/kernel/config/nvmet/hosts/` created at line 611.

Let me know what you think!

matteotenca commented:
https://github.com/pkoutoupis/rapiddisk/blob/1f668284f5b9341241268d39d13137475fa65934/src/nvmet.c#L541

It seems this array is not used anymore.

pkoutoupis commented:
I will move the unused array and for the nvmet mappings/unmappings...

**Scenario 1:**
That is desired behavior. Even if the host nqn path exists, it just skips it. But the idea is, if things change such as port no. exporting or whatever else, it will pass through to the section of the function.

**Scenario 2:** 
I will look at this tomorrow. Thanks for checking.

**Scenario 3:**
That is desired because another target may be mapped to the same host. I will just leave it there. A reboot will clean it up and it doesn't occupy any real space in memory to leave it there. In a future release, I may add a function to remove hosts as a separate command.

pkoutoupis commented:
@matteotenca I just noticed this in `src/Makefile`

```
.PHONY: clean
clean:
	rm -f *.d $(OBJS_TOOL_NDEBUG) $(OBJS_DAEMON_NDEBUG) $(OBJS_TOOL_DEBUG) $(OBJS_DAEMON_DEBUG) $(BIN_TOOL_NDEBUG) $(BIN_DAEMON_NDEBUG) $(BIN_TOOL_DEBUG) $(BIN_DAEMON_DEBUG)
    MAKECMDGOALS:=$(filter-out clean,$(MAKECMDGOALS)))
.PHONY: tools
```
What is going on here with MAKECMDGOALS and is the indentation wrong?

pkoutoupis commented:
Also, I figured out the problem with scenario 2:

When attr_allow_any_host is set to 1, it does not allow for host NQNs to be linked in the allowed_hosts folder. So, I needed to change the order of things and first set attr_allow_any_host to 0, THEN create the link mappings.

After you answer me about the MAKECMDGOALS, I will push a new commit with all notes addressed.

matteotenca commented:
> @matteotenca I just noticed this in `src/Makefile`
> 
> ```
> .PHONY: clean
> clean:
> 	rm -f *.d $(OBJS_TOOL_NDEBUG) $(OBJS_DAEMON_NDEBUG) $(OBJS_TOOL_DEBUG) $(OBJS_DAEMON_DEBUG) $(BIN_TOOL_NDEBUG) $(BIN_DAEMON_NDEBUG) $(BIN_TOOL_DEBUG) $(BIN_DAEMON_DEBUG)
>     MAKECMDGOALS:=$(filter-out clean,$(MAKECMDGOALS)))
> .PHONY: tools
> ```
> 
> What is going on here with MAKECMDGOALS and is the indentation wrong?

`Makefile`s format requires shell commands to be prepended by a tabulation, while `make` commands to be prepended by spaces. You can check it by changing the four spaces to a tab, this is what happens:

```bash
shub@ubuserver:~/rapiddisk/src$ make clean
rm -f *.d main_ndebug.o utils_ndebug.o json_ndebug.o rdsk_ndebug.o nvmet_ndebug.o sys_ndebug.o rapiddiskd_ndebug.o utils-server_ndebug.o json-server_ndebug.o nvmet-server_ndebug.o net_ndebug.o rdsk-server_ndebug.o sys-server_ndebug.o main_debug.o utils_debug.o json_debug.o nvmet_debug.o rdsk_debug.o sys_debug.o rapiddiskd_debug.o utils-server_debug.o json-server_debug.o nvmet-server_debug.o net_debug.o rdsk-server_debug.o sys-server_debug.o rapiddisk rapiddiskd rapiddisk_debug rapiddiskd_debug
MAKECMDGOALS:=)
/bin/sh: 1: Syntax error: ")" unexpected
make: *** [Makefile:81: clean] Error 2
shub@ubuserver:~/rapiddisk/src$
```

The second line is considered a shell command, so `sh` is invoked. Whith spaces, it is considered an internal `make` command.

Regards

pkoutoupis commented:
@matteotenca I just pushed a new commit (the date/time of my VM needs to be updated, so it shows up a couple of status updates above). Let me know what you think and if you give it the thumbs up, please change your review status.

matteotenca commented:
> @matteotenca I just pushed a new commit (the date/time of my VM needs to be updated, so it shows up a couple of status updates above). Let me know what you think and if you give it the thumbs up, please change your review status.

### Just and idea
In my very humble opinion, the exporting/unexporting process should be handled apart from the process of choosing allowed hosts for an export. I don't know if it is mandatory to specify the allowed hosts upon exporting, though.

I say this because the unexport function is quite tricky just because it has to manage the allowed host list too: when more than one host is present, the function behaviour switches to a bare handling of the access list, since it cannot unexport until the allowed hosts list is empty.

My opinion is the `-e` and `-x` should just trigger exporting and unexporting, and that other two switches should be added to choose allowed hosts for an exported target, like this fictional example:

```bash
$ sudo rapiddisk -e -b rd0 -P 1
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been mapped through port 1 as nqn.2021-06.org.rapiddisk:ubuserver-rd0.
$ sudo rapiddisk -A nqn.2021-06.org.rapiddisk:ubuserver-rd0 -H hostone
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Host 'hostone' is now allowed to access nqn.2021-06.org.rapiddisk:ubuserver-rd0.
$ sudo rapiddisk -D nqn.2021-06.org.rapiddisk:ubuserver-rd0 -H hostone
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Host 'hostone' is no more allowed to access nqn.2021-06.org.rapiddisk:ubuserver-rd0.
$
```

### Issues

Anyway, there are some issues regarding the messages printed to the user.

1. This call will cause a ugly message to be printed when the specified host is not in the allowed hosts list:
```bash
shub@ubuserver:/tmp/rapiddisk/src$ sudo ./rapiddisk_debug -x -b rd0 -P 1 -H foobar
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

nvmet_unexport_volume: Host foobar does not exist
shub@ubuserver:/tmp/rapiddisk/src$
```
2. The JSON output when removing an host when many hosts are allowed to access is broken.
In text mode is fine:
```bash
shub@ubuserver:/tmp/rapiddisk/src$ sudo ./rapiddisk_debug -x -b rd0 -P 1 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been unmapped from NVMe Target host test.
The target is still mapped to host NQN test4.
The target is still mapped to host NQN test3.
The target is still mapped to host NQN test2.
shub@ubuserver:/tmp/rapiddisk/src$
```
Same command but in JSON mode after command `sudo ./rapiddisk_debug -e -b rd0 -P 1 -H test`:
```bash
shub@ubuserver:/tmp/rapiddisk/src$ sudo ./rapiddisk_debug -e -b rd0 -P 1 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been mapped to test through port 1 as nqn.2021-06.org.rapiddisk:ubuserver-rd0.
shub@ubuserver:/tmp/rapiddisk/src$ sudo ./rapiddisk_debug -j -x -b rd0 -P 1 -H test | jq
{
  "status": "Success",
  "message": "The target is still mapped to host NQN test2."
}
shub@ubuserver:/tmp/rapiddisk/src$
```

3. There are other issues related to messages printed in console mode and not in JSON mode.

Since are all cosmetic changes, to speed up things, I corrected the code, attached thre is the diff file gzipped.

Regards!
[nvmet.c.diff.gz](https://github.com/pkoutoupis/rapiddisk/files/10377770/nvmet.c.diff.gz)

pkoutoupis commented:
@matteotenca Pushed your commit with minor changes (I forgot to change the date of my VM, again):
1. Removed an extra "." at the end of:
```
msg = "Block device %s could not be removed from the NVMe Target subsystem for unspecified host. %s";
```
2. And corrected wording for NVMe hosts. It shouldn't be NVMe Target Host but instead it should be NVMe Initiator Host. RapidDisk is the Target and the external hosts accessing it are the Initiators.

Thanks for catching that. It didn't dawn on me on how it would format in JSON. It is always helpful having another set of eyes and I do appreciate it. Again, let me know if this looks good and if so, please change your review status accordingly.

ALSO, per your idea / suggestion. We can talk it over https://github.com/pkoutoupis/rapiddisk/issues/148 and work on it for 9.1 or later.

matteotenca commented:
> Thanks for catching that. It didn't dawn on me on how it would format in JSON. It is always helpful having another set of eyes and I do appreciate it. 

As usual, I am glad my contribution is appreciated!

> Again, let me know if this looks good and if so, please change your review status accordingly.

I believe there is another glitch regarding exporting and unexporting operations, when no previous call to network interface enabling for NVMe Target exporting was accomplished.

Look at this:

```bash
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -e -b rd0 -P 1 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Error. Unable to create link of NQN to port. nvmet_export_volume: symlink: No such file or directory
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$
```
This makes sense, since there was not a previous command invokation using the `-i` argument.
But prior to exiting with that error, a lot of operations were already carried on on the filesystem: directory creations and wtiting to files. This happens because the existance of the port specified on the command line is never done.

This leads to a messy situation:

```bash
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        1: NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd0         Namespace: 1    Device: /dev/rd0        Enabled: True

Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$
```

And it becomes impossible to get rid of that export:

```bash
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -x -b rd0
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been removed from the NVMe Target subsystem.
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        1: NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd0         Namespace: 1    Device: /dev/rd0        Enabled: True

Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -x -b rd0 -P 1
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

nvmet_unexport_volume: Port 1 and / or export does not exist.
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        1: NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd0         Namespace: 1    Device: /dev/rd0        Enabled: True

Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -x -b rd0 -P 1 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Initiator Host 'test' does not exist.
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        1: NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd0         Namespace: 1    Device: /dev/rd0        Enabled: True

Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -x -b rd0 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Initiator Host 'test' does not exist.
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -x -b rd0
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been removed from the NVMe Target subsystem.
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        1: NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd0         Namespace: 1    Device: /dev/rd0        Enabled: True

Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$
```

I am pushing a commit and create a PR. A list of the ports is created and the specified port is checked against it, if not in the list, nothing is done. If port is `INVALID_VALUE` upon exporting, it is considered valid.

pkoutoupis commented:
@matteotenca Ping. Do you feel good for a merge / 9.0.0 tag as of the latest commit?

matteotenca commented:
> @matteotenca Ping. Do you feel good for a merge / 9.0.0 tag as of the latest commit?

I found some memory leaks and created a PR... but there is an issue with RDMA ports. When a RDMA port is created, the exporting operations fail, and the environment becomes "dirty". Since exporting and unexporting fails in different steps, the kernel filesystem situation ends up to be messed up. I believe that export/unexport functions should be modified so that they can undo the operations already carried on upon failure... :-\

Look at this sequence of commands:

```bash
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        None.


Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -N
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -i enp0s3 -P 1 -t rdma
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Successfully created port 1 set to rdma for interface enp0s3 (with IP address 192.168.20.40).
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -N
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Exported NVMe Ports

        1: Port: 1 - 192.168.20.40 (rdma)
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        None.


Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -e -b rd0 -P 1 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Error. Unable to create link of NQN to port. nvmet_export_volume: symlink: No such device
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        1: NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd0         Namespace: 1    Device: /dev/rd0        Enabled: True

Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -N
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Exported NVMe Ports

        1: Port: 1 - 192.168.20.40 (rdma)
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -x -b rd0 -P 1 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Block device rd0 has been unmapped from NVMe Initiator host test.
nvmet_unexport_volume: Port 1 and / or export does not exist.
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        1: NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd0         Namespace: 1    Device: /dev/rd0        Enabled: True

Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -N
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Exported NVMe Ports

        1: Port: 1 - 192.168.20.40 (rdma)
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -x -b rd0 -P 1 -H test
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Initiator Host 'test' does not exist.
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -x -b rd0 -P 1
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

nvmet_unexport_volume: Port 1 and / or export does not exist.
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        1: NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd0         Namespace: 1    Device: /dev/rd0        Enabled: True

Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -N
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Exported NVMe Ports

        1: Port: 1 - 192.168.20.40 (rdma)
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -X -P 1
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target port 1 has been removed.
shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        1: NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd0         Namespace: 1    Device: /dev/rd0        Enabled: True

Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk/src$ sudo ./rapiddisk_debug -N
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk/src$

```

pkoutoupis commented:
Ok. I merged your branch. Thank you for finding and fixing your discovery.

As with the second part of your comment:
It looks like the error is a result of the port not being configured for RDMA. dmesg states the following:
```
[ 3664.472868] nvmet_rdma: binding CM ID to 10.0.0.185:4420 failed (-19)
```
Which makes sense. In order to use RDMA over standard ethernet, you need to configure soft RoCE. Years ago, I wrote an article for Linux Journal to do exactly that: [https://www.linuxjournal.com/content/data-flash-part-ii-using-nvme-drives-and-creating-nvme-over-fabrics-network](https://www.linuxjournal.com/content/data-flash-part-ii-using-nvme-drives-and-creating-nvme-over-fabrics-network). Anyway, it seems like the process has changed since and instead of typing:
```
# rxe_cfg add eth0
```
You need to type:
```
# rdma link add rxe_eth0 type rxe netdev eth0
```
I just tested it and it worked for RDMA without error.
```
petros@dev-machine:~/rapiddisk/src$ sudo ./rapiddisk -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

	1: NQN: nqn.2021-06.org.rapiddisk:dev-machine-rd0 	Namespace: 1	Device: /dev/rd0 	Enabled: True

Exported NVMe Ports

	1: Port: 2 - 10.0.0.185 (tcp)	NQN: nqn.2021-06.org.rapiddisk:dev-machine-rd0
	2: Port: 1 - 10.0.0.185 (rdma)	NQN: nqn.2021-06.org.rapiddisk:dev-machine-rd0
```

So, to get back to your original point:
This will be better addressed in the NVMe target subsystem rework designated for https://github.com/pkoutoupis/rapiddisk/issues/148. I will add these notes there. Thank you for testing this.

Anyway, all of that aside, what do you think about merging now and as soon as we do, I can start working on separating the functions for #148?

alexmyczko commented:
how likely is the release to happen? https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1029266

pkoutoupis commented:
@alexmyczko Thank you for this information.

@matteotenca, thoughts?

matteotenca commented:
> @alexmyczko Thank you for this information.
> 
> @matteotenca, thoughts?

I believe we are ready to go. Issues we already are aware of can be fixed in a minor release.

pkoutoupis commented:
I will cut the new release tomorrow. Thank you!!!
